### ISSUE TITLE
Authentication with Keycloak keeps redirecting to `Login with SSO` in console

### ISSUE THREAD

**Issue by zhilongli** [minio/minio#13214]
I followed the instructions [here](https://github.com/minio/minio/blob/master/docs/sts/keycloak.md) to enable auth with keycloak for Minio. I was able to successfully run the `docs/sts/web-identity.go` script to see the buckets and policies associated with the user after logging in via the Keycloak login page. However, when I try to access the console for the first time, I get redirected to the Keycloak login page, and upon keying in the credentials I keep getting redirected back to the minio console login page with the `Login with SSO` button. 

## Expected Behavior
After authenticating with Keycloak, I should be able to access the Minio console dashboard. 

## Current Behavior

After authentication with Keycloak, I get stuck on the `Login with SSO` page. 

## Possible Solution

Not entirely sure.

## Steps to Reproduce (for bugs)

1. Install Minio server standalone on Ubuntu 20.04 following the instructions [here](https://github.com/minio/minio#gnulinux)
2. Configure keycloak client and user policy following the instructions [here](https://github.com/minio/minio/blob/master/docs/sts/keycloak.md#configure-keycloak-realm)
3. Install MC on Ubuntu 20.04, set the alias to the standalone server, and run 
```
mc admin config set myminio identity_openid config_url=<url> client_id=<id> client_secret=<secret> redirect_uri=<console url> &&
mc admin service restart myminio
```
4. Access the minio console to log in. 

## Context

Trying to set up SSO for Minio with Keycloak as the external identity provider. 

## Your Environment

* installed minio server from github master branch. 
* Ubuntu 20.04
* Tried installing minio server both on the server directly, and on a Kubernetes cluster on the server using a bitnami helm chart for minio, both ran into the same problem. 

Thanks for any help here!

**Comment by vadmeste** [comment]
`redirect_uri=<console url>`

Are you sure you put the address of the console with the correct port here ?

**Comment by harshavardhana** [comment]
@zhilongli make sure you have `policy` claim set properly in your JWT. You should see an error logged on the server side.

**Comment by zhilongli** [comment]
Thank you both for your response.

@vadmeste yes I'm certain that the redirect url specified for `mc admin config set` is the exact same url and port that I set for the Minio console. 

@harshavardhana I don't see any logs printed in my terminal running the minio server when I try to log in via SSO; and I believe that I have the `policy` claim set properly, since when I decode the JWT from running `docs/sts/web-identity.go` I can see the policy attribute for the user that I set within Keycloak. 

As an additional detail, I looked at the network tab in the browser to track the requests, and noticed that there's a 401 unauthorized error when it is trying to request to `<console_url>/api/v1/session` (see screenshot), before redirecting back to the `Login with SSO` page. 
![image](https://user-images.githubusercontent.com/36818454/133536495-05d88cee-b698-4a0b-971a-16a20d92e635.png)


Thanks again.

**Comment by harshavardhana** [comment]
> Thank you both for your response.
> 
> @vadmeste yes I'm certain that the redirect url specified for `mc admin config set` is the exact same url and port that I set for the Minio console. 
> 
> @harshavardhana I don't see any logs printed in my terminal running the minio server when I try to log in via SSO; and I believe that I have the `policy` claim set properly, since when I decode the JWT from running `docs/sts/web-identity.go` I can see the policy attribute for the user that I set within Keycloak. 
> 
> As an additional detail, I looked at the network tab in the browser to track the requests, and noticed that there's a 401 unauthorized error when it is trying to request to `<console_url>/api/v1/session` (see screenshot), before redirecting back to the `Login with SSO` page. 
> ![image](https://user-images.githubusercontent.com/36818454/133536495-05d88cee-b698-4a0b-971a-16a20d92e635.png)
> 
> 
> Thanks again. 

This policy needs to match a policy created on , MinIO using `mc admin policy add`

**Comment by zhilongli** [comment]
Yes, I assigned the predefined `readwrite` policy to the user, and I can see the `readwrite` policy when I run `mc admin policy list myminio`. I verified again by running the `docs/sts/web-identity.go` client and decoded the resultant `SessionToken` in jwt.io, below is a snippet of the decoded token. As can be seen in the token there is the `policy` field of `readwrite`

![image](https://user-images.githubusercontent.com/36818454/133618347-df1efdf6-2886-4796-acea-acbb46a9fa99.png)

**Comment by Scapal** [comment]
I'm running Minio (gateway nas) + Etcd + KeyCloak.

Upgrading the Minio Docker image from RELEASE.2021-08-25T00-41-18Z  to RELEASE.2021-09-24T00-24-24Z breaks the authentication. Every other settings remaining the same.

The "_Login With SSO_" is calling the Minio URL instead of the OpenID server.

On the 2021-08-25 version, the "_Login With SSO_" button has this link: `https://kc.mydomain.com/auth/realms/master/protocol/openid-connect/auth?client_id=minio-cstorage&redirect_uri=https%3A%2F%2Fcstorage.mydomain.com%3A9001%2Foauth_callback&response_type=code&scope=openid+address+web-origins+phone+profile+offline_access+email+microprofile-jwt+roles&state=TEFHUFlaM0dOMEwwNUxNQVRDEADBEEFDEADBEEFuRDRzK2gwWWFmWFkxTEpkSGdvPQ%3D%3D`

On the 2021-09-24 version: 
`https://cstorage.mydomain.com:9001/login?client_id=minio-cstorage&redirect_uri=https%3A%2F%2Fcstorage.mydomain.com%3A9001%2Foauth_callback&response_type=code&scope=openid+address+web-origins+phone+profile+offline_access+email+microprofile-jwt+roles+openid&state= TEFHUFlaM0dOMEwwNUxNQVRDEADBEEFDEADBEEFuRDRzK2gwWWFmWFkxTEpkSGdvPQ%3D%3D`

**Comment by Scapal** [comment]
When trying to log in, I can now see the KeyCloak server being called by the Minio server itself with `GET /auth/realms/master` but without any other parameter.
Is Minio now trying to proxy the request to the OpenID server instead of having the "_Login With SSO_" button sending to it directly ?

**Comment by zhilongli** [comment]
> I'm running Minio (gateway nas) + Etcd + KeyCloak.
> 
> Upgrading the Minio Docker image from RELEASE.2021-08-25T00-41-18Z to RELEASE.2021-09-24T00-24-24Z breaks the authentication. Every other settings remaining the same.
> 
> The "_Login With SSO_" is calling the Minio URL instead of the OpenID server.
> 
> On the 2021-08-25 version, the "_Login With SSO_" button has this link: `https://kc.mydomain.com/auth/realms/master/protocol/openid-connect/auth?client_id=minio-cstorage&redirect_uri=https%3A%2F%2Fcstorage.mydomain.com%3A9001%2Foauth_callback&response_type=code&scope=openid+address+web-origins+phone+profile+offline_access+email+microprofile-jwt+roles&state=TEFHUFlaM0dOMEwwNUxNQVRDEADBEEFDEADBEEFuRDRzK2gwWWFmWFkxTEpkSGdvPQ%3D%3D`
> 
> On the 2021-09-24 version:
> `https://cstorage.mydomain.com:9001/login?client_id=minio-cstorage&redirect_uri=https%3A%2F%2Fcstorage.mydomain.com%3A9001%2Foauth_callback&response_type=code&scope=openid+address+web-origins+phone+profile+offline_access+email+microprofile-jwt+roles+openid&state= TEFHUFlaM0dOMEwwNUxNQVRDEADBEEFDEADBEEFuRDRzK2gwWWFmWFkxTEpkSGdvPQ%3D%3D`

Yes I had the same issue where the latest release of Minio doesn't redirect to the Keycloak login page. Previous releases worked.

**Comment by zhilongli** [comment]
> I followed the instructions [here](https://github.com/minio/minio/blob/master/docs/sts/keycloak.md) to enable auth with keycloak for Minio. I was able to successfully run the `docs/sts/web-identity.go` script to see the buckets and policies associated with the user after logging in via the Keycloak login page. However, when I try to access the console for the first time, I get redirected to the Keycloak login page, and upon keying in the credentials I keep getting redirected back to the minio console login page with the `Login with SSO` button.
> 
> ## Expected Behavior
> After authenticating with Keycloak, I should be able to access the Minio console dashboard.
> 
> ## Current Behavior
> After authentication with Keycloak, I get stuck on the `Login with SSO` page.
> 
> ## Possible Solution
> Not entirely sure.
> 
> ## Steps to Reproduce (for bugs)
> 1. Install Minio server standalone on Ubuntu 20.04 following the instructions [here](https://github.com/minio/minio#gnulinux)
> 2. Configure keycloak client and user policy following the instructions [here](https://github.com/minio/minio/blob/master/docs/sts/keycloak.md#configure-keycloak-realm)
> 3. Install MC on Ubuntu 20.04, set the alias to the standalone server, and run
> 
> ```
> mc admin config set myminio identity_openid config_url=<url> client_id=<id> client_secret=<secret> redirect_uri=<console url> &&
> mc admin service restart myminio
> ```
> 
> 1. Access the minio console to log in.
> 
> ## Context
> Trying to set up SSO for Minio with Keycloak as the external identity provider.
> 
> ## Regression
> ## Your Environment
> * installed minio server from github master branch.
> * Ubuntu 20.04
> * Tried installing minio server both on the server directly, and on a Kubernetes cluster on the server using a bitnami helm chart for minio, both ran into the same problem.
> 
> Thanks for any help here!

Re: My original question of failed login after authenticating with Keycloak, it turned out that it was my realm settings that was interfering with the login, because I was reusing a realm that I had for another app. Starting from scratch following the instructions with a new realm worked.

**Comment by harshavardhana** [comment]
> When trying to log in, I can now see the KeyCloak server being called by the Minio server itself with `GET /auth/realms/master` but without any other parameter.

There was an issue introduced recently which is fixed in 84dcd25a36bdaa00ad4167282bc2ca19e6b29f26

**Comment by zhilongli** [comment]
I will close this issue since my original problem was solved. Thanks.
